sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala

```diff
@@ -658,6 +669,62 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     val intResultsB: Seq[Int] = Seq(314000000, 314200000, 314160000, 314159000, 314159300,
       314159260) ++ Seq.fill(7)(314159265)
 
+    def doubleResultsFloor(i: Int): Any = {
+      val results = Seq(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 3,
+        3.1, 3.14, 3.141, 3.1415, 3.14159, 3.141592)
+      if (i <= 6) results(i).toLong else results(i)
+    }
+
+    def doubleResultsCeil(i: Int): Any = {
+      val results = Seq(1000000.0, 100000.0, 10000.0, 1000.0, 100.0, 10.0,
+        4L, 3.2, 3.15, 3.142, 3.1416, 3.1416, 3.141593)
+      if (i <= 6) results(i).toLong else results(i)
+    }
+
+    def floatResultsFloor(i: Int): Any = {
+      val results = Seq(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 3L,
+        3.1d, 3.14d, 3.141d, 3.1414d, 3.14149d, 3.141499d)
+      if (i <= 6) results(i).toLong else results(i)
+    }
+
+    def floatResultsCeil(i: Int): Any = {
+      val results = Seq(1000000.0f, 100000.0f, 10000.0f, 1000.0f, 100.0f, 10.0f,
+        4L, 3.2d, 3.15d, 3.142d, 3.1415d, 3.1415d, 3.1415d)
+      if (i <= 6) results(i).toLong else results(i)
+    }
+
+    def shortResultsFloor(i: Int): Any = {
+      val results = Seq[Long](0L, 0L, 30000L, 31000L,
+        31400L, 31410L, 31415L) ++ Seq.fill[Long](7)(31415)
+      results(i)
+    }
+
+    def shortResultsCeil(i: Int): Any = {
```

Comment:
Since the data type of `ceil` is always double/long, I think we should only test double and long. It's not useful to test byte/short/int/float, as Spark will cast them to double/long and we don't need to test the cast here.

Reply:
I think we need to keep this to test `Ceil` with Scale parameters to verify whether we cast it to right type.

Reply:
This test should focus on expression evaluation and we should test `RoundCeil` diretly. If we want to test the added cast, we should add a new test case, to call `Ceil.apply` and check the result expression to see if it has the rigjt cast, without evaluating it.

Reply:
Sorry, i am not sure i got it right. 

Actually we test `RoundCeil` and `RoundFloor` directly for positive `scale` on big decimal, i can add long and double in that case, if you prefer. But we can't test `RoundCeil` and `RoundFloor` directly for negative `scale` because the casting to `Long` for negative scale happens one step before. 

May be if put the casting in the `RoundCeil` and `RoundFloor` objects and let the `ExpressionBuilder` call these objects like it was before, we can able to test `RoundCeil` and `RoundFloor` directly for both negative and positive `scale`. What do you say ?

Reply:
@cloud-fan could you please add your opinion on this point as well ? Do you agree to put the cast in the `RoundCeil`/`RoundFloor` object to be able to unit test the `RoundCeil` and `RoundFloor` directly for both +ve and -ve scale ?

Reply:
Ideally, expression builder should simply be a router and create corresponding expressions for certain input types. The type coercion shouldn't be done there. The created expression should define its expected input types and let the type coercion framework to add casts.

Reply:
Ok perfect, i'll put back the casting in the `RoundCeil` and `RoundFloor` objects then.